### ISSUE TITLE
Cmd-v to paste clipboard text in vterm throws buffer is read only error.

### ISSUE THREAD

**Issue by karthikbox** [akermu/emacs-libvterm#450]
I was trying out vtrerm in my emacs (terminal) and it works great until i realized that i'm not able to paste my clipboard contents into the vterm buffer.

Repro:
Cmd-c to copy some text 
Cmd-v to paste the text in the vterm buffer but i see "Buffer is read-only: #<buffer vterm>"

Am i missing something obvious here? 

I use emacs on a remote linux host via ssh.
My local machine is mac.

**Comment by Sbozzolo** [comment]
You should use Emacs kill/yank keybindings to copy&paste (e.g., to paste, you would use `M-y`). 
You can change this behavior by customizing your Emacs settings.

**Comment by karthikbox** [comment]
sbozzolo@ Thanks for the reply. My question was not about the emacs kill ring. Copy and paste between vterm and emacs killring works for me.

But my question is more about pasting mac clipboard text into the emacs vterm buffer.
Concretely, using Cmd-v, i'm able to paste text from my mac clipboard into any emacs buffer.
But when i try to do the same in a vterm buffer, i get the buffer read only error.

LMK if this makes sense, otherwise i'm happy to clarify.

**Comment by Sbozzolo** [comment]
I've never had a mac, so I don't know exactly how it behaves. On GNU/Linux, Emacs can read primary and clipboard from X, so I can paste into vterm buffers using Emacs' own functions. 

If anyone wants to help here, that would be great.

**Comment by jixiuf** [comment]
we have remap  `yank`  https://github.com/akermu/emacs-libvterm/blob/8d016d2d65d514cfa978f4541ce83b111007dd40/vterm.el#L548
and `s-v` runs the command yank (found in global-map) on macOS, so it should work.
Have you bind `s-v` to other command ?

**Comment by karthikbox** [comment]
In the vterm buffer of my terminal emacs (no x), s-v (or cmd-v) is bound to <xtrem-paste>. 
Should i remap this to vterm-yank?
` 
<xterm-paste> runs the command xterm-paste (found in global-map),     
   which is an interactive compiled Lisp function in ‘term/xterm.elc’.   
                                                                         
   It is bound to <xterm-paste>.                                         
                                                                        
   (xterm-paste EVENT)                                                   
                                                                        
   Handle the start of a terminal paste operation.
`

**Comment by halloleo** [comment]
I think in the vterm buffer you need to remap your paste key to `vterm-yank`, yes.

But check that the key really reaches Emacs (by typing in Emacs Ctrl-h k <yourkey>, then look at the help screen) and that not your terminal app does something with the key beforehand...

**Comment by jixiuf** [comment]
should be fixed .

**Comment by felker** [comment]
While the fix in 4743bfd does avoid the "buffer is read only" error, it does not result in the behavior I would expect. 

E.g. starting with a fresh Emacs, I copy "foo" to the macOS clipboard from Google Chrome with Cmd+c. Attempting to Cmd+v in a fresh Emacs session (empty kill ring) vterm buffer fails with "Kill ring is empty".

Whereas I can Cmd+v in `*scratch*` and "foo" appears in the buffer and is added to the `kill-ring`. Is there anyway to get the vterm buffer to behave like this?

In vterm buffer, `C-h k Cmd+v`:
```
<xterm-paste> runs the command vterm-yank (found in vterm-mode-map), which is
an interactive compiled Lisp function in ‘vterm.el’.

It is bound to C-y, s-v, <S-insertchar>, <S-insert>, <menu-bar> <edit> <paste>,
<xterm-paste>.

(vterm-yank &optional ARG)

Yank (paste) text in vterm.

Argument ARG is passed to ‘yank’.
```

In non-vterm buffer, `C-h k Cmd+v`:
```
<xterm-paste> runs the command xterm-paste (found in global-map), which is an
interactive compiled Lisp function in ‘term/xterm.el’.

It is bound to <xterm-paste>.

(xterm-paste EVENT)

Handle the start of a terminal paste operation.
```

**Comment by dispensable** [comment]
> While the fix in 4743bfd does avoid the "buffer is read only" error, it does not result in the behavior I would expect.

same problem here, I just can't paste from system clipboard to emacs kill-ring with C-shift-insert. Press these keys on vterm mode just send all the kill-ring history with a  `~ ` postfix to the vterm with a error msg "M-[ 2 0 1 is undefined". Really frustrating.

**Comment by arthurcgusmao** [comment]
I am experiencing this problem when trying to paste from a clipboard manager into vterm. I use CopyQ on Ubuntu 16.04. It seems that the function used to paste text from an external program (such as CopyQ) into an Emacs buffer uses a different function and would require the buffer to not be read-only. Could there be a workaround for this?
